src/utils/first_follow.cpp

```diff
@@ -0,0 +1,267 @@
+      path.pop_back();
+      return nullables[key];
+    }
+
+    for (const auto &rule : GetRulesForParent(augmented_grammar, key)) {
+      std::vector<std::string>::const_iterator symbol_itr;
+      for (symbol_itr = rule.GetEntities().begin(); symbol_itr != rule.GetEntities().end(); symbol_itr++) {
+        // exit loop if non-nullable symbol is encountered
+        if (!calc_recursive(*symbol_itr, path)) {
+          break;
+        }
+      }
+
+      // for production A -> X Y Z, A is nullable iff X, Y, Z all are nullable
+      if (symbol_itr == rule.GetEntities().end()) {
+        nullables[key] = true;
+        path.pop_back();
+        return true;
+      }
+    }
+
+    // by default, a symbol is considered non-nullable
+    nullables[key] = false;
+    path.pop_back();
+    return false;
+  };
+
+  for (const auto &production : augmented_grammar) {
+    auto path = std::vector<std::string>(0);
+    calc_recursive(production.GetParent(), path);
+  }
+
+  return nullables;
+}
+
+std::unordered_map<std::string, std::vector<std::string>> CalcFirsts(
+    const grammar::Productions &augmented_grammar, const std::unordered_map<std::string, bool> &nullables) {
+  std::unordered_map<std::string, std::vector<std::string>> firsts;
+  // finished -> used to check if any new symbols are added to any FIRST(non-terminal) in a particular iteration
+  // useful in case of cycles in productions
+  bool finished = false;
+  for (const auto &production : augmented_grammar) {
+    for (const auto &rules : production.GetRules()) {
+      for (const auto &symbol : rules.GetEntities()) {
+        if (!SearchParent(augmented_grammar, symbol)) {
+          // if symbol is not present as a parent in grammar, then it is a terminal
+          // if X is a terminal, then FIRST(X) = {X}
+          firsts.insert(make_pair(symbol, std::vector<std::string>(1, symbol)));
+        }
+      }
+    }
+  }
+
+  /* at this point, map of FIRST only contains terminal -> {terminal} mappings */
+  std::function<std::vector<std::string>(const std::string &, std::vector<std::string> &)> calc_recursive;
+  // key  -> current symbol being explored
+  // path -> list of all the symbols visited uptil reaching key
+  // lambda function that returns vector<string>
+  calc_recursive = [&](const std::string &key, std::vector<std::string> &path) {
+    // FIRST(terminal) = {terminal}
+    if (!SearchParent(augmented_grammar, key)) {
+      return firsts[key];
+    }
+
+    // if key "belongs to" path, then first of key has already been calculated / visited
+    if (find(path.begin(), path.end(), key) != path.end()) {
+      return firsts[key];
+    }
+
+    // marking current symbol as visited
+    path.push_back(key);
+
+    /* now only un-visited non-terminals are left to be explored */
+
+    // registering non-terminal in FIRST map if encountered for the first time
+    if (firsts.find(key) == firsts.end()) {
+      firsts.insert(make_pair(key, std::vector<std::string>(0)));
```

Comment:
same, why be specific about zero-length vector when that comes by default?

Reply:
```
firsts[key] = {};
```